File: python/22.括号生成.py
#
# @lc app=leetcode.cn id=22 lang=python3
#
# [22] 括号生成
#


# @lc code=start
class Solution:
    # 按括号序列的长度递归：看不懂
    @lru_cache(None)
    def generateParenthesis(self, n: int) -> List[str]:
        if n == 0:
            return ['']
        ans = []
        for c in range(n):
            for left in self.generateParenthesis(c):
                for right in self.generateParenthesis(n - c - 1):
                    ans.append('({}){}'.format(left, right))            
        return ans

    # 回溯：官解
    def generateParenthesis2(self, n: int) -> List[str]:
        ans = list()
        def backtrace(combination, left, right):
            if len(combination) == 2 * n:
                ans.append(''.join(combination))
                return
            if left < n:
                combination.append('(')
                backtrace(combination, left + 1, right)
                combination.pop()
            if right < left:
                combination.append(')')
                backtrace(combination, left, right + 1)
                combination.pop()
        backtrace([], 0, 0)
        return ans

    # 回溯：自解
    def generateParenthesis1(self, n: int) -> List[str]:
        counts = {
            '(': 0,
            ')': 0,
        }

        def backtrace(index):
            if index == n * 2:
                combinations.append(''.join(combination))
            else:
                if counts['('] < n:
                    counts['('] += 1
                    combination.append('(')
                    backtrace(index + 1)
                    combination.pop()
                    counts['('] -= 1
                if counts[')'] < counts['(']:
                    counts[')'] += 1
                    combination.append(')')
                    backtrace(index + 1)
                    combination.pop()
                    counts[')'] -= 1

        combination = list()
        combinations = list()
        backtrace(0)
        return combinations


# @lc code=end

File: cpp/22.括号生成.cpp
/*
 * @lc app=leetcode.cn id=22 lang=cpp
 *
 * [22] 括号生成
 */
#include <algorithm>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
// @lc code=start
class Solution
{
public:
    vector<string> generateParenthesis(int n)
    {
        vector<string> combinations;

        string combination;
        // backtrack1(combinations, combination, 0, n, 0);
        backtrack2(combinations, combination, 0, 0, n);

        return combinations;
    }

    // 官解
    void backtrack2(vector<string>& combinations, string& combination, int open,
                    int close, int n)
    {
        if (combination.size() == n * 2) {
            combinations.push_back(combination);
            return;
        }

        if (open < n) {
            combination.push_back('(');
            backtrack2(combinations, combination, open + 1, close, n);
            combination.pop_back();
        }

        if (close < open) {
            combination.push_back(')');
            backtrack2(combinations, combination, open, close + 1, n);
            combination.pop_back();
        }
    }

    // 自写
    void backtrack1(vector<string>& combinations, string& combination,
                    int index, int n, int remained)
    {
        if (index == n * 2) {
            if (remained == 0) {
                combinations.push_back(combination);
            }
            return;
        }

        combination.push_back('(');
        backtrack1(combinations, combination, index + 1, n, remained + 1);
        combination.pop_back();

        if (remained > 0) {
            combination.push_back(')');
            backtrack1(combinations, combination, index + 1, n, remained - 1);
            combination.pop_back();
        }
    }
};
// @lc code=end
